components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/graph/AuthenticationGraphConfig.java

```diff
@@ -0,0 +1,202 @@
+
+    /* User can incorrectly configure cyclic graph. This will eliminate the cyclic graph evaluation. */
+    private static final int MAX_RECURSION_COUNT = 32;
+
+    private Node startNode;
+    private Node endNode;
+    private String name;
+    private Map<String, Node> nodesMap;
+
+    /**
+     * Builds the graph with Axiom.
+     * @param graphOM OM element of the graph.
+     * @return the built graph.
+     */
+    public static AuthenticationGraphConfig build(OMElement graphOM) {
+
+        AuthenticationGraphConfig authenticationGraphConfig = null;
+
+        Map<String, Node> nodesMap = new HashMap<>();
+        Iterator iterator = graphOM.getChildElements();
+        while (iterator.hasNext()) {
+            Object o = iterator.next();
+            if (o instanceof OMElement) {
+                OMElement nodeOM = (OMElement) o;
+                switch (nodeOM.getLocalName()) {
+                case "AuthenticationStep":
+                    StepNode stepNode = toStepNode(nodeOM);
+                    nodesMap.put(stepNode.getName(), stepNode);
+                    break;
+                case "AuthenticationDecision":
+                    DecisionNode decisionNode = toDecisionNode(nodeOM);
+                    nodesMap.put(decisionNode.getName(), decisionNode);
+                    break;
+                default:
+                    log.error("Unknown element:" + nodeOM.getLocalName() + " in graph node");
+                }
+            }
+        }
+
+        String startNodeName = graphOM.getAttributeValue(new QName(null, "start"));
+
+        Node startNode = nodesMap.get(startNodeName);
+        if (startNode == null) {
+            log.error("There was no node found as the graph start node: " + startNodeName);
+        }
+
+        Map<Node, VisitedNodeInfo> visitedNodesInfo = new HashMap<>();
+        try {
+            visit(nodesMap, startNode, visitedNodesInfo);
+            authenticationGraphConfig = new AuthenticationGraphConfig();
+            authenticationGraphConfig.startNode = startNode;
+            authenticationGraphConfig.nodesMap = nodesMap;
+        } catch (IdentityApplicationManagementException e) {
+            log.error("Error in building authentication graph", e);
+        }
+
+        return authenticationGraphConfig;
+    }
+
+    public Node getNodeByName(String name) {
+        return nodesMap.get(name);
+    }
+
+    private static void visit(Map<String, Node> nodesMap, Node currentNode,
+            Map<Node, VisitedNodeInfo> visitedNodesInfoMap) throws IdentityApplicationManagementException {
+        VisitedNodeInfo visitedNodeInfo = visitedNodesInfoMap.get(currentNode);
+        if (visitedNodeInfo == null) {
+            visitedNodeInfo = new VisitedNodeInfo();
+            visitedNodesInfoMap.put(currentNode, visitedNodeInfo);
+        }
+        visitedNodeInfo.incrementReferenceCount();
+        if (visitedNodeInfo.getReferenceCount() > MAX_RECURSION_COUNT) {
```

Comment:
+1, Need to have a reference counter per each node. Skip the node processing if the refCounter >0. This will allow cycles in the graph. Graph execution should also validate if the runtime goes to same path in a loop and should have a bailout if it is the case.